### ISSUE TITLE
ability to request admin privileges

### ISSUE THREAD

**Issue by cdmistman** [nushell/nushell#9649]
### Related problem

I'm writing a nu command that requires admin privileges, but I can't figure out a way to re-run the command via `sudo` to ensure admin privileges.

### Describe the solution you'd like

perhaps a command `request admin` which prompts the user to allow admin privileges of the current-running process

### Describe alternatives you've considered

i've tried something like this, but it fails with `sudo: foo: command not found`:
```nu
def main [] {
  if (is-admin) {
    print 'ok!'
  } else {
    sudo main
  }
}
```

afaict there's also no equivalent to `$0` so the script can't be re-invoked either

### Additional context and details

a rough equivalent in a `sh` script would be:
```sh
if [ "$USER" != 'root' ]; then sudo "$0"; fi
```

**Comment by amtoine** [comment]
what about just asking the user to run the command again with `sudo`, e.g. with `sudo !!`? :yum:

**Comment by cdmistman** [comment]
yes, that works and is what i already do, nonetheless it's better UX to be able to auto-upgrade, especially if the command can only ever run with privileges

nu might even be able to innovate here, this is an idea i had (though i know nothing of the feasibility)
- `request admin (error_struct)` will pause execution of the current scope
- the user is prompted for authentication with a clear indicator that the command/script is requesting admin privileges
- if the user denies privileges and `error_struct` is present, execution halts and the caller execution returns `false` (ie, `main` in the original issue description returns `false`)
- if the user denies privileges and `error_struct` is absent, then `request admin` returns `false`
- if the user grants privileges, then the current scope continues execution with said privileges

**Comment by amtoine** [comment]
i've never seen a command rerunning itself with `sudo` inside :thinking: 
i'm not sure i'd like that as a user :confused: 

the only ones i can think are `systemctl` and `paru` (an ArchLinux AUR helper) which prompts for a `sudo` password when not given.

also, i'm not sure that kind of feature will be the focus of the core team on the road to 1.0...
HOWEVER, we are not closed to contributions, more the opposite actually!
so i guess if someone can come up with a simple and safe implementation of the *specification* you give above, we could definitely consider reviewing it and merging it :blush:

**Comment by qdm12** [comment]
On Windows, running a command as administrator is rather cumbersome. For example with Powershell I would do `Start-Process $Program -Verb RunAs -Wait -PassThru -ArgumentList $Arguments`. I guess I can do some ugly `pwsh Start-Process $Program -Verb RunAs -Wait -PassThru -ArgumentList $Arguments` in nushell, but a `run-as-admin` command working cross platform would be awesome indeed.

**Comment by umnikos** [comment]
I'd like to revive interest in this issue because I tried to reimplement `sudo su -c "echo 50 > /sys/class/backlight/intel_backlight/brightness"` in nushell (without resorting to running the same command with bash from nu) and it was not easy:
```
~/do/fi/functions (main ↑4 +1) $ echo 50 | sudo -sE save -a /sys/class/backlight/intel_backlight/brightness
Error: nu::shell::cant_convert

  × Can't convert to string.
   ╭─[source:1:1]
 1 │ save -a \/sys\/class\/backlight\/intel_backlight\/brightness
   · ──┬─
   ·   ╰── can't convert nothing to string
   ╰────

~/do/fi/functions (main ↑4 +1) $ sudo -sE "echo 50 | save -a /sys/class/backlight/intel_backlight/brightness"
Error: nu::shell::external_command

  × External command failed
   ╭─[source:1:1]
 1 │ echo\ 50\ \|\ save\ -a\ \/sys\/class\/backlight\/intel_backlight\/brightness
   · ──┬──
   ·   ╰── did you mean 'echo'?
   ╰────
  help: No such file or directory (os error 2)

~/do/fi/functions (main ↑4 +1) $ def a [] { echo 50 | save -a /sys/class/backlight/intel_backlight/brightness }
~/do/fi/functions (main ↑4 +1) $ sudo -sE a 
Error: nu::shell::external_command

  × External command failed
   ╭─[source:1:1]
 1 │ a
   · ┬
   · ╰── executable was not found
   ╰────
  help: No such file or directory (os error 2)

~/do/fi/functions (main ↑4 +1) $ sudo nu -c "echo 50 | save -a /sys/class/backlight/intel_backlight/brightness"
sudo: nu: command not found 
~/do/fi/functions (main ↑4 +1) $ whereis nu
nu: /home/umnikos/.cargo/bin/nu                                                        
~/do/fi/functions (main ↑4 +1) $ sudo (^which nu) -c "echo 50 | save -a /sys/class/backlight/intel_backlight/brightness"
```
Passing any arguments at all through `sudo` seems to be a big challenge as they all have to be turned into strings, so a way to just run a closure as root would be appreciated. Maybe `do --as-root`?

**Comment by cdmistman** [comment]
since writing to files isn't the only operation that's common in scripts like these, it's probably better to have broad support for `sudo` inside scripts rather than selective support for only `save`

**Comment by 132ikl** [comment]
this should work:

```
def main [] {
  if (is-admin) {
    print 'ok!'
  } else {
    sudo $nu.current-exe $env.CURRENT_FILE
  }
}
```

let me know if this solves your issue 😄

**Comment by 132ikl** [comment]
hi, I'm going to close this issue. if you don't think this has been resolved, let me know and I'll re-open the issue.

**Comment by IndigoLily** [comment]
@132ikl how am I supposed to pass arguments and flags with this solution?
